src/control/tools/StrokeStabilizer.cpp

```diff
@@ -0,0 +1,354 @@
+}
+
+void StrokeStabilizer::Arithmetic::averageAndPaint(const Event& ev, guint32 timestamp) {
+    /**
+     * Push the event and overwrite the oldest event in the buffer
+     */
+    eventBuffer.push_front(ev);
+
+    /**
+     * Average the coordinates using an arithmetic mean
+     */
+    Event sum = std::accumulate(begin(eventBuffer), end(eventBuffer), Event(0, 0, 0), [](auto&& lhs, auto&& rhs) {
+        return Event(lhs.x + rhs.x, lhs.y + rhs.y, lhs.pressure + rhs.pressure);
+    });
+
+    /**
+     * Rescale the averaged coordinates and draw
+     */
+    double d = (double)eventBuffer.size();
+    sum.pressure /= d;
+    sum.x /= d;
+    sum.y /= d;
+    setLastPaintedEvent(sum);
+    drawEvent(sum);
+}
+
+auto StrokeStabilizer::Arithmetic::getLastEvent() -> Event {
+    if (eventBuffer.empty()) {
+        g_warning("StrokeStabilizer::Arithmetic buffer empty. This should never be!");
+        return Event(0, 0, 0);
+    }
+    return eventBuffer.front();
+}
+
+void StrokeStabilizer::Arithmetic::resetBuffer(Event& ev, guint32 timestamp) {
+    if (eventBuffer.back() != ev) {
+        eventBuffer.assign(bufferLength, ev);  // Empty and refill the buffer with copies of ev
+    }
+}
+
+/**
+ * StrokeStabilizer::VelocityGaussian
+ */
+void StrokeStabilizer::VelocityGaussian::recordFirstEvent(const PositionInputData& pos) {
+    eventBuffer.emplace_front(pos);
+    lastEventTimestamp = pos.timestamp;
+}
+
+void StrokeStabilizer::VelocityGaussian::averageAndPaint(const Event& ev, guint32 timestamp) {
+
+    /**
+     * Compute the velocity (if possible) and push the event to eventBuffer
+     */
+    if (eventBuffer.empty()) {
+        eventBuffer.emplace_front(ev);
+    } else {
+        /**
+         * Issue: timestamps are in ms. They are not precise enough. Different events can have the same timestamp.
+         */
+        VelocityEvent& last = eventBuffer.front();
+        guint32 timelaps = timestamp - lastEventTimestamp;
+        if (timelaps == 0) {
+            timelaps = 1;
+        }
+        eventBuffer.emplace_front(ev, hypot(ev.x - last.x, ev.y - last.y) / ((double)timelaps));
+    }
+    lastEventTimestamp = timestamp;
+
+    /**
+     * Average the coordinates using the gimp-like weights
+     */
+    Event weightedSum = {0, 0, 0};
+    double weight;
+    double sumOfWeights = 0;
+    double sumOfVelocities = 0;
+
+    auto it = eventBuffer.cbegin();
+    for (; it != eventBuffer.cend(); ++it) {
+        /**
+         * The first weight is always 1
+         */
+        weight = exp(-sumOfVelocities * sumOfVelocities / twoSigmaSquared);
+        if (weight < 0.01) {
+            break;
+        }
+        sumOfVelocities += (*it).velocity;
+        weightedSum.x += weight * (*it).x;
+        weightedSum.y += weight * (*it).y;
+        weightedSum.pressure += weight * (*it).pressure;
+        sumOfWeights += weight;
+    }
+    eventBuffer.erase(it, eventBuffer.cend());
```

Comment:
I don't know how to do that with a range loop (i.e. remember the breaking point). If someone knows, I'll comply to the conventions ;-)

Reply:
You could try to achieve the same using a `std::find_if`. But that would imply, that you are searching for something.
Use `std::find_if` is you are searching something / an iterator by predicate.
Use `std::for_each` when using prepared functors or if you have iterator ranges. 
Use range-based loop, when nothing of the above apply. (Will often be the case)
Try to apply `<algorithm>` functions like `std::accumulate`.
If nothing else seems to fit well your solution is the best what you can do.
Avoid always index based loops.

Reply:
This loop does enough work to be left as a simple for loop. It would not be easier to express this as a composition of different looping primitives.

Reply:
IMO, each of Febbe's solutions would do only half of the job: I could do the averaging using an `std::accumulate`, I could find the first event with neglectable weight with an `std::find_if`, but those operations need to be intertwined. Splitting them into two different loops would be a waste. 

Probably using a `std::find_if` with a weird lambda computing the sum and the weight as the `UnitaryPredicament` is possible. I fail to see the benefit of that: for someone reading the code, it makes it more mysterious than it should be.